athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/domain/spill/SpillLocationVerifier.java

```diff
@@ -0,0 +1,115 @@
+package com.amazonaws.athena.connector.lambda.domain.spill;
+
+/*-
+ * #%L
+ * Amazon Athena Query Federation SDK
+ * %%
+ * Copyright (C) 2019 - 2020 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to track the bucket and its state, and check its validity
+ */
+public class SpillLocationVerifier
+{
+    private static final Logger logger = LoggerFactory.getLogger(SpillLocationVerifier.class);
+
+    private enum BucketState
+    {UNCHECKED, VALID, INVALID}
+
+    private final AmazonS3 s3;
+    private String bucket;
+    private BucketState state;
+
+    /**
+     * @param bucket The name of the spill bucket.
+     */
+    public SpillLocationVerifier(AmazonS3 s3, String bucket)
+    {
+        this.s3 = s3;
+        this.bucket = bucket;
+        this.state = BucketState.UNCHECKED;
+    }
+
+    /**
+     * Public function used to check if the account that calls the lambda function owns the spill bucket
+     *
+     * @param spillBucket The name of the spill bucket.
+     */
+    public void checkBucketAuthZ(String spillBucket)
+    {
+        if (spillBucket == null) {
+            throw new RuntimeException("Spill bucket cannot be null!");
```

Comment:
This shouldn't be a runtime exception, the bucket may be empty if the customer is not using spill.  null or empty bucket should cause you to return.

Reply:
I don't understand the purpose of letting the costumer setting this to null or empty when they don't need spill. I thought we implement bucket auth checking to guard against user error. In this case, the customer could pass in null or empty bucket accidentally without the intend of bypassing spill. My point is, why do we give the customer a free pass when they think their query will not spill (when it might).

Reply:
Not all customers want to use spill, but when they do we want to ensure they are spilling to a location they own.

Reply:
I think I'm starting to understand your point (as I was writing a rebuttal...:-)). Bucket ownership is a high risk issue, a customer's data privacy is paramount, we'd rather be defensive, so we check ownership when the customer put in a spill bucket. But if they leave the spill bucket field empty (empty String) or remove the spill bucket field in Lambda (null pointer), they choose not to use a spill bucket. If their query requires spilling, the worst that can happen is that Lambda will throw a bucket unauthz exception from the S3Spiller and there is no data security issue.

I hope we are on the same page?

Reply:
not sure I 100% follow but if they don't set a spill bucket then we wont spill so there can't be a violation of the semantics this class is trying to enforce. This class is enforcing that you only write to spill buckets you own. it isn't trying to enforce that you have a spill bucket set. :)

Reply:
I agree, SpillLocationVerifier should handle if the customer owns the bucket and nothing more. But as far as I know, SDK still spills even when the customer doesn't provide a spill bucket. The logic in the SDK is: if (size of data > in-memory capacity) spill happens, and the customer doesn't have control over the config for in-memory capacity. So I just want to point out that if the customer provides no spill bucket, spilling might still happen, and at that point, the SDK may throw an S3 exception.

Reply:
I did some tests. In current configuration, if SpillLocationVerifier doesn't throw exception when spillbucket is null or empty, Athena console will show the following error: 

Your query has the following error(s):

GENERIC_INTERNAL_ERROR: BucketName cannot be empty/null

The SDK logs in CloudWatch went through without any warnings, and there was no spill. So even if the customer enter null or empty spillbucket, Athena still throws an error somewhere.